### ISSUE TITLE
Cannot use variables with where clause

### ISSUE THREAD

**Issue by hydrosIII** [nushell/nushell#13994]
### Describe the bug

Using a variable in a where clause with =! returns all the table and does not filter anything. 

### How to reproduce

Code:
Buggy code
```
let table_example = [[a b]; [1 2] [3,null]]  
let variable = "b"
$table_example |  where $"($variable)" != null

```
Returns
![image](https://github.com/user-attachments/assets/a3621acb-f8bf-48af-b88f-638b2874eefb)

### Expected behavior

Should return:

![image](https://github.com/user-attachments/assets/cab17dc8-f8b6-4217-9afe-3987c50ee0ff)

It is working without variable substitution:
```
let table_example = [[a b]; [1 2] [3,null]]    
$table_example |  where "b" != null


```

### Configuration

| key                | value                               |
| ------------------ | ----------------------------------- |
| version            | 0.98.0                              |
| major              | 0                                   |
| minor              | 98                                  |
| patch              | 0                                   |
| branch             |                                     |
| commit_hash        |                                     |
| build_os           | linux-x86_64                        |
| build_target       | x86_64-unknown-linux-gnu            |
| rust_version       | rustc 1.81.0 (eeb90cda1 2024-09-04) |
| rust_channel       | stable-x86_64-unknown-linux-gnu     |
| cargo_version      | cargo 1.81.0 (2dbb1af80 2024-08-20) |
| build_time         | 2024-09-18 18:17:09 +02:00          |
| build_rust_channel | release                             |
| allocator          | mimalloc                            |
| features           | default, sqlite, trash              |
| installed_plugins  |                                     |

**Comment by NotTheDr01ds** [comment]
Not at the keyboard to test, but I'm guessing the variable-interpolated string is being compared as a string, where hard-coding the `b` is being used as a cell-path.  Check out the `into cell-path` command and see if that helps.

**Comment by NotTheDr01ds** [comment]
Yes (back at keyboard) - That's what it is. The "b" is simply a string in your failing example, so `"b" != null` is true for all rows.

You'll need to use the closure form:

```nu
let table_example = [[ a   b    ];
                     [ 1   2    ]
                     [ 3   null ]]

let variable = "b"
let cellpath = [ $variable ] | into cell-path

$table_example | where {|row|
  ($row | get $cellpath)  != null
}
```

Alternate form - A bit more concise, but perhaps too terse to be readable:

```nu
let table_example = [[ a   b    ];
                     [ 1   2    ]
                     [ 3   null ]]

let variable = "b"
let cellpath = [ $variable ] | into cell-path

$table_example | where { get $cellpath | $in != null }
```

**Comment by fdncred** [comment]
I think this will work without the cellpath stuff too.
```nushell
❯ $table_example | where { get $variable | $in != null }
╭─#─┬─a─┬─b─╮
│ 0 │ 1 │ 2 │
╰───┴───┴───╯
```

**Comment by NotTheDr01ds** [comment]
Much simpler - Yes.  Just needs to be in a closure to handle the pipeline.

**Comment by weirdan** [comment]
```nushell
$table_example | where ($it | get $variable) != null
```
is even simpler.

**Comment by hydrosIII** [comment]
Ok so from what I understand. The cellpath is a different type of variable than the string type? Because I tried describe for both and it was both string.

`
 "b" | describe 
`

gives as a result string 
and 

`
let variable = "b"
$variable | describe 
`

gives string also.

So, shouldn't they be equivalent in all cases? Otherwise is somewhat confusing. And it means variable substitution depends on other programs like where in this case.  And it is not reliable.

The kindly provided solutions are working but the sintax logic in them gives the result that: 

`
let variable = "b" 
$variable | describe 
`
gives string 
and
 
`
let cellpath = [ $variable ] | into cell-path     
$cellpath | describe 
`
 gives cellpath. 

So. Ok they are different types of data for different type of cases. That is clear. But then the fist statement of 

`
let table_example = [[a b]; [1 2] [3,null]]    
$table_example |  where "b" != null
`
Should just work when "b" is not string but cellpath. But in this example is is string. 

But the fact that

"b" 

and

let $variable = "b"

Have in fact different results is shocking, at least to me. 
Because at the same time this somehow implies that :

$variable is not "b", at least not fully. Which contradicts the "=" in the sintax 
This will make the statement of variable being "b" not true. 

This is contradictory.

**Comment by hydrosIII** [comment]
I think this a syntax bug. It makes variable substitution for strings difficult to use as it has a lot of corner cases. But I understand if the language is not focused on that.

**Comment by NotTheDr01ds** [comment]
Ah, apologies for missing the notification on your reply.  At least, I assume there was a GitHub app notification - I found it earlier but can't find it now!

I think there are two aspects to your question above.

### Where row-conditions

First, your original question (and most of the follow-up), centers around the `where` command.  The `where` command is very unique.  [From the help](https://www.nushell.sh/commands/docs/where.html), it can use a "shorthand" form called a "row condition".  I'm not even sure this syntax is well (or at all) documented, but it is *very* basic.  As you've noticed above, anything more than a simple comparison will require a closure.

(And *this* (I think) is the key to answering your question) -- As far as I can tell, the row-condition has to start with a *string literal* (yes, the string *type* here) which is used to refer to a column/field within the row.  Even starting with a parenthesis will cause the row-condition to become invalid.

It's important to realize that in a row-condition, it's not (again, as far as I can tell) in any-way-shape-or-form a `cell-path`.  It's a `string` literal *only*.  Anything else gets handled as a "normal" comparison.

So as a string literal, `where b != null` works. However, as an *interpolated string*, `where $"(variable)" != null` doesn't work (it's always true).

Other "non-string-literals" that would "fail" in various ways here:

* `ls | where true` (or `false`):  Returns all (or no) rows, always, so is useless
* Any form which always evaluates to `true` or `false`, e.g.:
  * `let v = "name"; ls | where $"($v)" =~ '^a'`: The same as your root problem - Always false.

* `ls | enumerate | flatten | where ("index" mod 2) == 0`:  Starts with a parenthesis, so not a valid row-expression and will fail at parse-time.

On the other hand, `ls | enumerate | flatten | where "index" mod 2 == 0` works - `index` is a string literal and the expression can be evaluated as a row-condition.

#### This could be handled in one of (at least) three ways:

* Document the row-condition syntax and limitations, which should be done regardless.
* Refuse an interpolated string here as the LHS, because (since it's not a string *literal*), it's going to be compared to the RHS as a *string*, not a row/field name.
* And there's certainly a case that could be made that row-conditions should be able to handle an interpolated string as a row-name.

The second and third options there are (I believe) certainly feasible, but they would require someone who felt strongly enough about the issue to submit a PR to implement either.

The problem is that the script-side solution is so easy - Switch to using closure-form instead of row-condition-form.  (Personal opinion), it doesn't seem like it would be worth someone's time to worry about changing the row-condition handling.  But I could be surprised!

### `into-cellpath`

There's a side question above about:

```nu
let table_example = [[a b]; [1 2] [3,null]]

# Doesn't work:
$table_example | where {  get ("b" | into cellpath) | $in != null }

# Works:
$table_example | where { get (["b"]| into cell-path) | $in != null }
```

I agree that's a bit odd, and I've had that thought before.

I think it's entirely more likely that someone would implement a request to *"Allow `into cell-path` to accept a string"*.  I've thought about looking at the code to see how hard that would be.

To me, it's an ergonomic issue.  I think it's completely natural to try what you tried (because I have as well) ... `"b" | into cell-path`, and it's surprising when it doesn't work.  Again, easy solution, but here to me it's the *logical* thing that I think (most) everyone tries first.  It's also likely (for me at least) to be a mistake that will be repeated.

And I have a feeling it's a pretty easy fix, but again, I haven't had a chance to check yet.

**Comment by NotTheDr01ds** [comment]
Adding this separately since I didn't want it to get lost in the long answer above.

I think you should (if you want), open two enhancement requests:

* Allow row-conditions to accept an interpolated variable on the LHS (or alternatively, error if attempted)
* Allow `into cell-path` to accept a string (column/key) or an int (row/index)

**Comment by WindSoilder** [comment]
Closing the issue because the question is answered.
